external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java

```diff
@@ -0,0 +1,298 @@
+    public static final int DEFAULT_MAX_UNCOMMITTED_OFFSETS = 10_000;    // 10,000 records
+
+    // Kafka property names
+    public interface Consumer {
+        String GROUP_ID = "group.id";
+        String BOOTSTRAP_SERVERS = "bootstrap.servers";
+        String ENABLE_AUTO_COMMIT = "enable.auto.commit";
+        String AUTO_COMMIT_INTERVAL_MS = "auto.commit.interval.ms";
+        String KEY_DESERIALIZER = "key.deserializer";
+        String VALUE_DESERIALIZER = "value.deserializer";
+    }
+
+    /**
+     * The offset used by the Kafka spout in the first poll to Kafka broker. The choice of this parameter will
+     * affect the number of consumer records returned in the first poll. By default this parameter is set to UNCOMMITTED_EARLIEST. <br/><br/>
+     * The allowed values are EARLIEST, LATEST, UNCOMMITTED_EARLIEST, UNCOMMITTED_LATEST. <br/>
+     * <ul>
+     * <li>EARLIEST means that the kafka spout polls records starting in the first offset of the partition, regardless of previous commits</li>
+     * <li>LATEST means that the kafka spout polls records with offsets greater than the last offset in the partition, regardless of previous commits</li>
+     * <li>UNCOMMITTED_EARLIEST means that the kafka spout polls records from the last committed offset, if any.
+     * If no offset has been committed, it behaves as EARLIEST.</li>
+     * <li>UNCOMMITTED_LATEST means that the kafka spout polls records from the last committed offset, if any.
+     * If no offset has been committed, it behaves as LATEST.</li>
+     * </ul>
+     * */
+    public enum FirstPollOffsetStrategy {
+        EARLIEST,
+        LATEST,
+        UNCOMMITTED_EARLIEST,
+        UNCOMMITTED_LATEST }
+
+    /**
+     * Defines when to poll the next batch of records from Kafka. The choice of this parameter will affect throughput and the memory
+     * footprint of the Kafka spout. The allowed values are STREAM and BATCH. STREAM will likely have higher throughput and use more memory
+     * (it stores in memory the entire KafkaRecord,including data). BATCH will likely have less throughput but also use less memory.
+     * The BATCH behavior is similar to the behavior of the previous Kafka Spout. De default value is STREAM.
+     * <ul>
+     *     <li>STREAM Every periodic call to nextTuple polls a new batch of records from Kafka as long as the maxUncommittedOffsets
+     *     threshold has not yet been reached. When the threshold his reached, no more records are polled until enough offsets have been
+     *     committed, such that the number of pending offsets is less than maxUncommittedOffsets. See {@link Builder#setMaxUncommittedOffsets(int)}
+     *     </li>
+     *     <li>BATCH Only polls a new batch of records from kafka once all the records that came in the previous poll have been acked.</li>
+     * </ul>
+     */
+    public enum PollStrategy {
+        STREAM,
+        BATCH
```

Comment:
@harshach You are combining two different features/issues here.  The first one is limiting the maximum number of replay.  +1 this is a great feature.  The second one is how we do a replay on a failure when we have not hit the replay limit yet.

This code keeps the entire tuple in memory along with its offset so when a failure happens we can replay the message without talking to kafka.  The original KafkaSpout would only keep the offset in memory and if it needed to replay the message it would go back to kafka, fetch an entire batch of messages and drop the ones that didn't need to be replayed.

My opinion is that failure is not the common case, so lets not optimize for it.  The current code is much simpler, but also much more memory intensive, so lets keep the in-memory replay for now, but file a JIRA and go back to reduce the memory usage by adding in the more complex code that uses less memory.

If the entire point of the BATCH config is to reduce the memory footprint of the spout, and at the same time make the throughput really bad, lets not do it.  Instead I would prefer to release a memory hungry spout sooner and spend our efforts on this second JIRA that would reduce the memory usage without impacting the common case throughput.  If there is something else about the BATCH config that I am missing, or is going to be important in the future please let me know.